lib/modules/apostrophe-schemas/public/js/array-modal.js

```diff
@@ -174,51 +254,86 @@ apos.define('apostrophe-array-editor-modal', {
           self.editItem();
         }
       });
+    };
 
-      self.$el.parents('[data-modal]').find('[data-apos-save-array-items]').on('click', function(e){
-        var $self = $(this);
-        self.saveItemState(function(){
-          self.saveArray();
-        });
-      });
-    }
+    // This method removes the item with the specified `id` property from the
+    // array. 
 
-    self.remove = function(_id) {
+    self.remove = function(id) {
       self.arrayItems = _.filter(self.arrayItems, function(choice) {
-        return choice.id !== _id;
+        return choice.id !== id;
       });
       return self.refresh();
     };
 
     self.refreshing = 0;
     self.last = [];
 
+    // Refresh (re-render) the list of items, then invoke the `onChange` method
+    // if they differ from the previous set. This method is debounced. If calls to
+    // this method are nested only two refreshes will take place: the initial one
+    // and also the last one, to ensure the impact of any changes made in nested
+    // function calls is seen. As a further optimization, only the last one actually
+    // updates the markup in the browser.
+    //
+    // If a callback is passed, it is always invoked, even if this
+    // refresh is being skipped as an optimization due to nesting.
+
     self.refresh = function(callback) {
       if (self.refreshing) {
         self.refreshing++;
-        return;
+        return callback && callback();
       }
       self.refreshing++;
       self.$arrayItems.html('');
       return self.html('arrayItems', { arrayItems: self.arrayItems, field: self.field }, function(html) {
-        self.$arrayItems.html(html);
+        // If calls are nested make sure only the last one actually updates the
+        // markup, as a visual optimization
+        if (self.refreshing === 1) {
+          self.$arrayItems.html(html);
+        }
         self.decrementRefreshing();
         var compare = JSON.stringify(self.arrayItems);
         if (self.last !== compare) {
           self.last = compare;
-          //self.onChange();
-        }
-        if (callback){
-          return callback();
+          self.onChange();
         }
+        return callback && callback();
       }, function(err) {
         self.decrementRefreshing();
-        if (callback){
-          return callback();
-        }
+        return callback && callback();
       });
 
     };
+    
+    // Invoked when the contents of the array have changed, after
+    // a refresh of the display. Invokes the limit mechanism.
+
+    self.onChange = function() {
+      self.limit();
+    };
+    
+    // Implements the `limit` option by showing and hiding
+    // the limit message and the add button, respectively.
+
+    self.limit = function() {
+      if (!self.field.limit) {
+        return;
+      }
+      if (self.arrayItems.length >= self.field.limit) {
+        self.$el.find('[data-apos-add-array-item]').hide();
```

Comment:
Would prefer to manage visual state by toggling a selector here. Not the end of the world and probably breaks patterns already being used.